File: framework/languages/java/e2e/PubSub/Server/Publisher/src/main/java/systems/zlink/e2e/pubsub/publisher/PublisherApplication.java
package systems.zlink.e2e.pubsub.publisher;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.boot.WebApplicationType;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.context.annotation.Bean;
import systems.zlink.e2e.pubsub.publisher.Configuration.PublisherOptions;
import systems.zlink.e2e.pubsub.publisher.Endpoints.PublisherEndpoints;
import systems.zlink.e2e.pubsub.publisher.Infrastructure.EvidenceStore;
import systems.zlink.e2e.pubsub.shared.Contracts;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationOptions;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;

@EnableZLinkFramework
@SpringBootApplication(proxyBeanMethods = false)
public final class PublisherApplication {
    public AutoCloseable run(String... args) {
        SpringApplicationBuilder builder = new SpringApplicationBuilder(PublisherApplication.class)
            .web(WebApplicationType.NONE);
        builder.application().setKeepAlive(true);
        return builder.run(args)::close;
    }

    @Bean
    PublisherOptions publisherOptions() {
        return PublisherOptions.fromEnv();
    }

    @Bean
    ObjectMapper objectMapper() {
        return new ObjectMapper();
    }

    @Bean
    EvidenceStore evidenceStore() {
        return new EvidenceStore();
    }

    @Bean
    PublisherEndpoints publisherEndpoints(
        PublisherOptions options,
        systems.zlink.framework.channels.ZLinkFanoutClient fanout,
        EvidenceStore evidence,
        ObjectMapper json) {
        return new PublisherEndpoints(options, fanout, evidence, json);
    }

    @Bean
    ZLinkFrameworkConfigurer publisherFramework(PublisherOptions options) {
        return framework -> {
            framework.configureDispatch()
                .messageFlow(ZLinkMessageFlowLogMode.KEY_TRANSITIONS)
                .traceLogFile(options.logDir() + "/publisher-flow.log")
                .traceLabel("java-ps-publisher");
            framework.addFanoutChannel(Contracts.EVENT_CHANNEL)
                .enablePublisher(options.publisherEndpoint());
        };
    }

    @Bean
    ZLinkRedisLocationStore locationStore(PublisherOptions options) {
        return new ZLinkRedisLocationStore(new ZLinkRedisLocationOptions()
            .setConnectionString(options.redisLocationEndpoint())
            .setKeyPrefix(options.locationKeyPrefix()));
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Publisher/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/publisher/PublisherApplication.kt
package systems.zlink.e2e.kotlin.spotservice.publisher

import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.annotation.Bean
import systems.zlink.contracts.core.RoutingId
import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.spots.ZLinkSpotPublisherClient
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer

@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackages = ["systems.zlink.e2e.kotlin.spotservice.publisher"]
)
class PublisherApplication {
    @Bean
    fun publisherFramework(): ZLinkFrameworkConfigurer =
        ZLinkFrameworkConfigurer { options ->
            val logDir = Env.get("ZLINK_KOTLIN_E2E_LOG_DIR", "logs")
            options.codecs().addJson()
            options.useDiscovery().addRegistryEndpoint(Env.get("ZLINK_KOTLIN_E2E_REGISTRY_ROUTER"))
            options.configureDispatch()
                .messageFlow(ZLinkMessageFlowLogMode.KEY_TRANSITIONS)
                .traceLogFile("$logDir/publisher-flow.log")
                .traceLabel("kotlin-sm-publisher")
            options.addSpotMesh(Contracts.SPOT_MESH)
                .enablePubSub(Env.get("ZLINK_KOTLIN_E2E_SPOT_PUBLISHER_ENDPOINT"))
                .setRoutingId(RoutingId.from("publisher"))
        }

    companion object {
        @JvmStatic
        fun run(vararg args: String) {
            val context = SpringApplicationBuilder(PublisherApplication::class.java)
                .web(WebApplicationType.NONE)
                .run(*args)
            try {
                val publisher = context.getBean(ZLinkSpotPublisherClient::class.java)
                publisher.publishSpot(
                    Contracts.SPOT_MESH,
                    "spot.events",
                    Contracts.MeshEvent("c4-publisher")
                )
                    .packetName("MeshEvent")
                    .await()
                println("scenario SM-C4 passed")
            } finally {
                context.close()
            }
        }
    }
}
